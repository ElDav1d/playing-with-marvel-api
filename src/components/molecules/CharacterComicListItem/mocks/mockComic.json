{
  "id": 21171,
  "dates": [
    {
      "type": "onsaleDate",
      "date": "2012-07-04T00:00:00-0400"
    },
    {
      "type": "focDate",
      "date": "2012-06-20T00:00:00-0400"
    },
    {
      "type": "unlimitedDate",
      "date": "2013-04-08T00:00:00-0400"
    },
    {
      "type": "digitalPurchaseDate",
      "date": "2012-07-04T00:00:00-0400"
    }
  ],
  "images": [
    {
      "path": "http://i.annihil.us/u/prod/marvel/i/mg/6/60/5ba3d0869c543",
      "extension": "jpg"
    },
    {
      "path": "http://i.annihil.us/u/prod/marvel/i/mg/c/a0/4fda5398edc41",
      "extension": "jpg"
    },
    {
      "path": "http://i.annihil.us/u/prod/marvel/i/mg/9/b0/4fda539915bf3",
      "extension": "jpg"
    },
    {
      "path": "http://i.annihil.us/u/prod/marvel/i/mg/6/d0/4fda5398a0685",
      "extension": "jpg"
    }
  ],
  "title": "Hulk (2008) #55",
  "description": "The hands of the doomsday clock race towards MAYAN RULE! Former Avengers arrive to help stop the end of the world as more Mayan gods return. Rick \"A-Bomb\" Jones falls in battle!",
  "issueNumber": 55,
  "modified": "2012-06-19T11:48:47-0400"
}
