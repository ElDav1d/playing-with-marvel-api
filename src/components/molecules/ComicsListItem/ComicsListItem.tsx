import { CharacterComicDetails } from '@/components/pages/CharacterDetail/interfaces/characterComics';
import Image from '@/components/atoms/Image';
import { useMemo } from 'react';
import { formatDate } from '@/utils/helpers';

export interface ComicsListItemProps {
  comic: CharacterComicDetails;
}

const ComicsListItem = ({ comic }: ComicsListItemProps) => {
  const dateTypes = ['onsaleDate', 'focDate'];

  const filteredDates = useMemo(() => {
    return comic?.dates?.filter((date) => dateTypes.includes(date.type));
  }, [comic]);

  return (
    <li>
      {comic.images.length > 0 && (
        <Image
          title={comic.title}
          alt={`The pic of ${comic.title}'s cover`}
          path={comic.images[0].path}
          extension={comic.images[0].extension}
          sizing='standard_small'
        />
      )}
      <h3>{comic.title}</h3>
      <p>{comic.description}</p>
      <p>
        <small>
          <strong>Issue: </strong>
          {comic.issueNumber}
        </small>
      </p>
      <p>
        <small>
          <strong>Modified: </strong>
          {formatDate(comic.modified)}
        </small>
      </p>
      {filteredDates.map((date) => (
        <p key={date.date}>
          <small>
            <strong>{date.type}: </strong>
            {formatDate(date.date)}
          </small>
        </p>
      ))}
    </li>
  );
};

export default ComicsListItem;
