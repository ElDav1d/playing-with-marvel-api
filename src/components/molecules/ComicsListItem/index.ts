export { default as ComicsListItem } from './ComicsListItem';
