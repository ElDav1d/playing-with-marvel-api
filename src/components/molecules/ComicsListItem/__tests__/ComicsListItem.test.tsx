import mockComic from '../mocks/mockComic.json';
import { render, screen } from '@testing-library/react';
import ComicsListItem, { IComicsListItemProps } from '../ComicsListItem';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

const queryClient = new QueryClient();

const comic: IComicsListItemProps = JSON.parse(JSON.stringify(mockComic));

it('renders a comic list item', () => {
  // ARRANGE
  render(
    <QueryClientProvider client={queryClient}>
      <ComicsListItem
        id={comic.id}
        images={comic.images}
        title={comic.title}
        description={comic.description}
      />
    </QueryClientProvider>,
  );

  // ASSERT
  screen.getByRole('listitem', { name: /hulk/i });
  screen.getByRole('heading', { name: /hulk/i });
});

it('matches snapshot', () => {
  // ARRANGE
  const { asFragment } = render(
    <QueryClientProvider client={queryClient}>
      <ComicsListItem
        id={comic.id}
        images={comic.images}
        title={comic.title}
        description={comic.description}
      />
    </QueryClientProvider>,
  );

  // ASSERT
  expect(asFragment()).toMatchSnapshot();
});
