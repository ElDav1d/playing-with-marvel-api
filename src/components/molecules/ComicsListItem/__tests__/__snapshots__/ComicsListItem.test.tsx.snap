// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <li
    aria-label="Hulk (2008) #55"
    class="bg-black text-white overflow-hidden relative z-0
    after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
  "
  >
    <picture
      class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
    />
    <div
      class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-s before:duration-l before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
    >
      <h3
        class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-3"
      >
        Hulk (2008) #55
      </h3>
      <p
        class="text-xs leading-5 line-clamp-3"
      >
        The hands of the doomsday clock race towards MAYAN RULE! Former Avengers arrive to help stop the end of the world as more Mayan gods return. Rick "A-Bomb" Jones falls in battle!
      </p>
    </div>
  </li>
</DocumentFragment>
`;
